pandas/_libs/lib.pyx

```diff
@@ -800,8 +801,30 @@ cpdef ndarray[object] ensure_string_array(
         # short-circuit, all elements are str
         return result
 
+    if arr.dtype.kind == "f":  # non-optimized path
+        for i in range(n):
+            val = arr[i]
```

Comment:
Personally, I'd rather spell this `for i, val in enumerate(arr):`. Probably worth comparing if this gives the same performance.